### PR TITLE
Support batch processing for deep neural networks

### LINKED ISSUE #141: Support batch processing of DNN
In current implementation of `dolphin-dnn`, the neural network model is trained with only one input instance at each time. If we train the model with input batches, we can get better performance on computing activation values especially in fully connected layers since activations of multiple input instances are computed with one matrix multiplication.
Also, in model partitioning, we can split input batches, and each evaluator in the same model takes these splits and computes activation values and gradients.

### PR DESCRIPTION
**Pull request by beomyeol**
This pull request introduces batch processing for deep neural networks. A batch size that is used for training can be specified in the neural network configuration file. The changes involved in this PR are the followings.
* An input batch is given for training a neural network or inferring outputs.
* The signature of `push` for `ParameterProvider` is changed. The size of batch should be given.
* The return matrix for `generateParameterGradient` is the element-wise sum of parameter gradient matrices for each input instance in a batch.
* The messages that are used for reducing parameter gradients with group communication are changed from a list of layer parameters to a list of pairs of an integer (the size of batch) and a layer parameter (the element-wise sum of parameter gradients for an input batch).

This closes #141.

### CODE REVIEW

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/IntAndLayerParameterArrayPairCodec.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Override
+  public byte[] encode(final Pair<Integer, LayerParameter[]> integerAndLayerParametersPair) {
+    try (final ByteArrayOutputStream bstream = new ByteArrayOutputStream();
+         final DataOutputStream dstream = new DataOutputStream(bstream)) {
+      encodeToStream(integerAndLayerParametersPair, dstream);
+      return bstream.toByteArray();
+    } catch (final IOException e) {
+      throw new RuntimeException("IOException during IntegerAndLayerParameterArrayPairCodec.encode()", e);
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/IntAndLayerParameterArrayPairCodec.java]: The class name is different: `IntAndLayerParameterArrayPairCodec` / `IntegerAndLayerParameterArrayPairCodec`?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/IntAndLayerParameterArrayPairListCodec.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @Override
+  public byte[] encode(final List<Pair<Integer, LayerParameter[]>> intAndLayerParametersPairList) {
+    try (final ByteArrayOutputStream bstream = new ByteArrayOutputStream();
+         final DataOutputStream dstream = new DataOutputStream(bstream)) {
+      encodeToStream(intAndLayerParametersPairList, dstream);
+      return bstream.toByteArray();
+    } catch (final IOException e) {
+      throw new RuntimeException("IOException during IntegerAndLayerParameterArrayPairListCodec.encode()", e);
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/IntAndLayerParameterArrayPairListCodec.java]: The class name is different: `IntAndLayerParameterArrayPairListCodec ` / `IntegerAndLayerParameterArrayPairListCodec `?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java [hunk 3]
```diff
@@ -83,13 +90,33 @@ public void parse() {
...
+        }
       } catch (final IOException e) {
         parseException = new ParseException("IOException: " + e.toString());
         return;
       }
     }
-    result = trainingData;
+
+    if (validationBatchGenerator.size() > 0) {
+      dataList.add(new Pair<>(validationBatchGenerator.pull(), false));
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: Shouldn't the boolean value be `true`?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java [hunk 4]
```diff
@@ -83,13 +90,33 @@ public void parse() {
         final Matrix data = input.get(range(0, input.getColumns() - 2));
         final int label = (int) input.get(input.getColumns() - 2);
         final boolean isValidation = ((int) input.get(input.getColumns() - 1) == 1);
-        trainingData.add(new Pair<>(new Pair<>(data, label), isValidation));
+
+        if (isValidation) {
+          final Pair<Matrix, int[]> batch = validationBatchGenerator.push(data, label);
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: Making `push` return a value seems a bit awkward. Does `BatchGenerator` have to be separated from this `NeuralNetworkDataParser` class, like the current code? Wouldn't be it okay to make `BatchGenerator` directly add pairs to `dataList` when the batch size is reached? The `matrixFactory` field of  `NeuralNetworkDataParser` could directly be used, too.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java [hunk 5]
```diff
@@ -39,26 +41,29 @@
...
-  private List<Pair<Pair<Matrix, Integer>, Boolean>> result;
+  private final int batchSize;
+  private List<Pair<Pair<Matrix, int[]>, Boolean>> result;
   private ParseException parseException;
 
   @Inject
   private NeuralNetworkDataParser(final MatrixFactory matrixFactory,
                                   final DataSet<LongWritable, Text> dataSet,
-                                  @Parameter(Delimiter.class)final String delimiter) {
+                                  @Parameter(Delimiter.class)final String delimiter,
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: We should've caught this before: whitespace between `.class)` and `final`.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java [hunk 6]
```diff
@@ -67,23 +56,17 @@ private GroupCommParameterProvider(
...
-      } catch (final NetworkException e) {
-        throw new RuntimeException("NetworkException while trying to send reduce", e);
-      } catch (final InterruptedException e) {
-        throw new RuntimeException("InterruptedException while trying to send reduce", e);
+  public synchronized void push(final int batchSize, final LayerParameter[] parameterGradients) {
+    try {
+      if (batchSize == 0 || parameterGradients == null || parameterGradients.length == 0) {
+        parameterGradientReduceSender.send(Collections.<Pair<Integer, LayerParameter[]>>emptyList());
+      } else {
+        parameterGradientReduceSender.send(Collections.singletonList(new Pair<>(batchSize, parameterGradients)));
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java]: The group comm version now uses `List<Pair>`s as Reduce values, but the compute tasks send lists that contain only one object. Instead of sending `List<Pair>`s, how about sending the inner `Pair`s directly? This way, the summation of parameter gradients can be done across reducers instead of only at the controller task.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetworkParameterUpdater.java [hunk 7]
```diff
@@ -103,47 +100,24 @@ public NeuralNetParamServerData process(final String key,
...
-   * Aggregate parameter gradients by computing the average of all gradients, per layer,
-   * and multiplying the step size.
+   * Multiplies the step size (in-place update).
    */
-  private List<LayerParameter[]> processLayerParametersList(final String key,
-                                                            final List<LayerParameter[]> parameterGradientsList) {
-    if (parameterGradientsList == null || parameterGradientsList.size() == 0 || !key.equals(WHOLE_MODEL)) {
+  private LayerParameter[] processLayerParameters(final String key,
+                                                  final LayerParameter[] parameterGradients) {
+    if (parameterGradients == null || parameterGradients.length == 0 || !key.equals(WHOLE_MODEL)) {
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetworkParameterUpdater.java]: Could we take out `parameterGradients.length == 0`? I know it's an unexpected condition, but its meaning is completely different from `parameterGradientsList.size() == 0`. Or, should we treat it just like any other error?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java [hunk 8]
```diff
@@ -113,4 +140,75 @@ public void parse() {
...
+    private Matrix makeBatch(final List<Matrix> inputs) {
+      if (inputs.size() > 0) {
+        final Matrix ret = matrixFactory.create(inputs.size(), inputs.get(0).getLength());
+        int i = 0;
+        for (final Matrix vector : inputs) {
+          ret.putRow(i++, vector);
+        }
+        return ret;
+      } else {
+        throw new IllegalArgumentException("At least one ndarray is needed to make batch");
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: Perhaps, "At least one **matrix**"?

**Comment by jsjason** [comment]
@beomyeol Sorry for the late pass. I've left some comments for you. The sample dataset works well on both groupcomm and paramserver configurations.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java (hunk 4 above)
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: As you mentioned, it will be some benefits not to have separate class for generating batches. However, without `BatchGenerator`, a code for data and label lists for training and validation inputs, and handling these lists, will look too complicated. That's why I introduced `BatchGenerator` to make code more clear.
Anyway, I will modify `push` method that returns a value and make `BatchGenerator` not static in order to use `matrixFactory` and `batchSize` directly.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetworkParameterUpdater.java (hunk 7 above)
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetworkParameterUpdater.java]: I'll remove a code for checking `parameterGradients.length == 0` at `processLayerParameters()` and add an error handling code for zero-length parameter gradients array at `updateLayerParameter()`.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java (hunk 6 above)
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java]: I tried to send `Pair` instead of `List<Pair>`. But, I encountered problems about how to define an empty message for synchronization and how a reducer handle this empty message for aggregation.

If we define `<0, null>` as an empty message, we should treat a `null` array at codecs.
If we define `<0, LayerParameter[0]>` as an empty message, there is an inconsistency for the number of layers and the number of layer parameters. (but, we may ignore this inconsistency.)
For `<0, LayerParameter[n]>` with `null` elements, we should treat `null` elements of an array at codecs.

With `List<Pair>`, the empty message is clear and easy to handle it at reducers. If you have a better idea, please let me know :)

**Comment by beomyeol** [comment]
@jsjason Thank you for your review. I've addressed your comments and left a comment for the type of reduce values. Please take another look :)

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java (hunk 6 above)
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java]: In my opinion, the performance boost that can be gained by using `Pair`s and applying summation at reduce functions worths more than the code clearness we can achieve by using `List<Pair>`s. I think `<0, LayerParameter[0]>` is best for an empty message, because we can just ignore the length of the layer parameter array entirely if the batch size is zero.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java [hunk 9]
```diff
@@ -83,13 +90,33 @@ public void parse() {
         final Matrix data = input.get(range(0, input.getColumns() - 2));
         final int label = (int) input.get(input.getColumns() - 2);
         final boolean isValidation = ((int) input.get(input.getColumns() - 1) == 1);
-        trainingData.add(new Pair<>(new Pair<>(data, label), isValidation));
+
+        if (isValidation) {
+          validationBatchGenerator.push(data, label);
+          if (validationBatchGenerator.check()) {
```
Comment (jsjason) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java]: I'm completely fine with having the inner class `BatchGenerator`, but I still think this code can be improved to be more concise. `check()` could be called inside `push`, and the next line `dataList.add(new Pair...)` could also be run inside `BatchGenerator` instead of here. I don't quite see why that would make the code more complicated.

**Comment by jsjason** [comment]
@beomyeol I've done a second pass and left two comments. Thanks!

**Comment by beomyeol** [comment]
@jsjason I've addressed your comments. Please review the changes.

**Comment by jsjason** [comment]
Looks great, will merge.
